multicluster wrap add enhance roundtripper, support informer cache

champly opened the pull request:
Signed-off-by: champly <champly@outlook.com>

`multicluster.NewClusterGatewayRoundTripper` not support informer cache. Because informerFactory.Start use ctx listen for stop signals. So add enhance rounttripper include clusterName.

msftgits commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/oam-dev/cluster-gateway?pullRequest=35) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/oam-dev/cluster-gateway?pullRequest=35) before we can accept your contribution.<br/><br/>:x: champly [sign now](https://cla.opensource.microsoft.com/oam-dev/cluster-gateway?pullRequest=35)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/oam-dev/cluster-gateway?pullRequest=35) it.</sub>

champly commented:
@yue9944882 PTAL

yue9944882 reviewed (commented):
thanks for the contribution. the original example in this package is for showing the capability of dynamically switching between the clusters (by passing a context attached the target cluster name). 

while i see the intention of ur new example is to instantiate a new informer __dedicated for a single cluster__ over the cluster gateway, so it'd better to place ur new examples a new folder probably named `single-cluster-informer`. meanwhile for clarification we can also renamed the original example to `dynamic-multicluster-client`. and in long terms, ideally there should be a dynamic multi-cluster informer prior to the proposing single cluster informer.

last but not least, i had a complex feeling about providing such utilities over cluster-gateway b/c generally it's not recommended to run an informer (from the hub cluster) to list-watch the managed clusters. this is a legacy pattern practiced in the kubefed v2 project, which is proved to have a bottleneck in performance when the # of clusters grows. it's recommended to follow the "pull" pattern i.e. running an agent from the managed clusters to "pull" tasks from the hub so the informers should be running in the managed cluster. on the other hand, i heard other voices looking for the capabilities of list-watching-from-hub so it may be wise to leave the door open.

champly commented:
> thanks for the contribution. the original example in this package is for showing the capability of dynamically switching between the clusters (by passing a context attached the target cluster name).
> 
> while i see the intention of ur new example is to instantiate a new informer **dedicated for a single cluster** over the cluster gateway, so it'd better to place ur new examples a new folder probably named `single-cluster-informer`. meanwhile for clarification we can also renamed the original example to `dynamic-multicluster-client`. and in long terms, ideally there should be a dynamic multi-cluster informer prior to the proposing single cluster informer.
> 
> last but not least, i had a complex feeling about providing such utilities over cluster-gateway b/c generally it's not recommended to run an informer (from the hub cluster) to list-watch the managed clusters. this is a legacy pattern practiced in the kubefed v2 project, which is proved to have a bottleneck in performance when the # of clusters grows. it's recommended to follow the "pull" pattern i.e. running an agent from the managed clusters to "pull" tasks from the hub so the informers should be running in the managed cluster. on the other hand, i heard other voices looking for the capabilities of list-watching-from-hub so it may be wise to leave the door open.

Thanks you for your suggestion. I also agree with your point of view. In fact, we don't use informer cache, but we need to mark `managedcluster`, so we need a single cluster instead of dynamic. I have modified the file structure, PTAL.

yue9944882 commented on examples/single-cluster-informer/main.go:
```diff
@@ -0,0 +1,91 @@
+
+			mgr, err := controllers.NewManager(cfg, manager.Options{Scheme: s})
+
+			podInformer2, err := mgr.GetCache().GetInformer(context.TODO(), &corev1.Pod{})
```
nit: avoid numbers in the variable name, in favor of `runtimePodInformer`

yue9944882 commented on pkg/apis/cluster/transport/enroundtripper.go:
```diff
@@ -0,0 +1,30 @@
+
+var _ EnhanceClusterGatewayRoundTripper = &enhanceClusterGatewayRoundTripper{}
+
+type EnhanceClusterGatewayRoundTripper interface {
```
"enhance" is an ambiguity name. in favor of a name will give users an intuitive sense of what the utilitity is doing. fwiw, "ProxyPathPrependingClusterGatewayRoundTripper" might be better?

yue9944882 reviewed (commented):
a few nits in naming, otherwise LGTM

yue9944882 reviewed (commented):
LGTM
